clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -136,11 +139,55 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
         if (s == null) {
             s = new Sensor(this, name, parents, config == null ? this.config : config, time);
             this.sensors.put(name, s);
+            if (parents != null) {
+                for (Sensor parent : parents) {
+                    List<Sensor> children = childrenSensors.get(parent.name());
+                    if (children == null) {
+                        children = new ArrayList<>();
+                        childrenSensors.put(parent.name(), children);
+                    }
+                    children.add(s);
+                }
+            }
         }
         return s;
     }
 
     /**
+     * Remove a sensor (if it exists), associated metrics and its children.
+     *
+     * @param name The name of the sensor to be removed
+     */
+    public void removeSensor(String name) {
+        List<Sensor> removedSensors = new ArrayList<>();
+
+        synchronized (this) {
+            removeSensor(name, removedSensors);
+        }
+
+        /*
+         * Because `Sensor.metrics` acquires a lock on `sensor` and `Sensor.add` calls methods on `Metrics` that
+         * acquire a lock on the metrics instance, it is important to do the `sensor.metrics` calls below outside
+         * the `synchronized(this)` above.
+         */
+        for (Sensor sensor : removedSensors)
+            for (KafkaMetric metric : sensor.metrics())
+                removeMetric(metric.metricName());
+    }
+
```

Comment:
Perhaps we can lock the Sensor first and then lock the base metrics instance whenever we are removing a sensor. During addMetric, the sensor is locked first and then the Metrics object. If we follow the same pattern when doing removeSensor, we should not have any deadlocks right?

What do you think? I'm still open to handling this in KAFKA-2419, but I want to be sure we have an acceptable solution.